llvm/lib/Transforms/Yk/ShadowStack.cpp

```diff
@@ -1,11 +1,85 @@
+// memory is then stored into a global variable for other functions to pick up.
+//
+// Then for each non-main function, F, we then insert a "shadow prologue" which:
+//  1. load's the shadow stack's "high water mark" from the global variable.
+//  2. adds to the pointer to reserve shadow space for F.
+//  3. stores the new high water mark back to the global variable.
+//
+// Then for each non-main function, at each return point from the function,
+// we insert code to restore the shadow stack pointer back to what it was
+// before we adjusted it.
+//
+// If a function requires no shadow space, then the above steps can be omitted.
+//
+// main() is assumed to not recursively call main(), as this would cause use to
+// re-allocate the shadow stack from scratch, leaking the existing shadow stack
+// and generally causing chaos. This is checked during this pass.
+//
+// Special considerations regarding setjmp/longjmp:
+//
+// Consider a function using setjmp like this:
+//
+// ```
+// define f() {
+//   ; allocate shadow space
+//   %0 = load ptr, ptr @shadowstack_0, align 8
+//   %1 = getelementptr i8, ptr %0, i32 16
+//   store ptr %1, ptr @shadowstack_0, align 8
+//   ...
+//   call @setjmp(...)
+//   ...
+//   ; and suppose g() uses the shadow stack and calls longjmp() to transfer to
+//   ; the above setjmp().
+//   call @g(...)
+//   ...
+//   call @h(...) ; assume this also uses the shadow stack.
+//   ...
+//   return:
+//    ; release shadow space
+//    store ptr %0, ptr @shadowstack_0, align 8
+//    ret i32 1
+// }
+// ```
+//
+// Is the system in a consistent after g() calls longjump()?
+//
+// This question can be split into two:
+//
+// 1. Will f() restore the shadow stack pointer to the right value when it
+// returns after the longjmp()?
+//
+// 2. Will callees like h() get a useable shadow stack after the longjmp()?
+//
+// To answer 1: since %0 is local to f() and not changed between the calls the
+// setjmp and longjmp, this value will be restored during longjmp() and thus
+// the epilogue will do the right thing. See the man page for setjmp(3) for
+// more on why this works.
+//
+// To answer 2: when g() longjumps, we will skip g()'s shadow epilogue, so the
+// shadow stack pointer will not be restored. In effect, g()'s shadow frame
+// leaks up until the point where f()'s shadow epilogue restores the shadow
+// stack pointer.
+//
+// Does this matter? I believe the behaviour to be correct in the sense that
+// h() will get a usable shadow frame, but just deeper in the shadow stack than
+// expected. One can dream up scenarios, e.g. "h(), calls i(), calls j(), ...,
+// which longjumps", where a long chain of shadow frames temporarily leak,
+// potentially blowing the shadow stack. This could happen if the top-level
+// interpreter loop (which is long-lived) contains a commonly jumped-to
+// setjump(), so we probably do want to fix this soon.
+//
+// YKFIXME: If you wanted to fix this you'd have to reload the shadow stack
```

Comment:
This is what I wanted to discuss.

I wonder if we should try to fix this right off the bat. That's assuming you agree with the reasoning outlined in the comment.

It requires a little bit of implementation effort, but should be doable.

Reply:
Right, so I think what this PR does is already "correct" but (at least temporarily) can leak arbitrary amounts of shadow stack space. I'm fine with that temporarily: I think the fix for the leak is for `setjmp` (not `longjmp!`) to be nobbled in the way Lukas is suggesting.

Reply:
> and put it right after the setjump call.

my idea was to reload the shadow pointer immediately before and then update all later uses to the reload, but in hindsight, I think Lukas' solution is simpler to implement.

Reply:
> but (at least temporarily) can leak arbitrary amounts of shadow stack space

Indeed. That's what I concluded too.

Reply:
> my idea was to reload the shadow pointer immediately before and then update all later uses to the reload

I think that would only reset the local shadow stack pointer, but as you correctly pointed out in case 1) this is already done automatically by the long jump. Storing that value back into the shadow stack global ensures subsequent calls also load the correct pointer value from the global.

Reply:
I don't think you can easily load it back before you jump: you don't know where the right value will be in the `jmpbuf`. But, if you could do that, it would I think be equivalent to doing it after the `setjmp`. Either way, the basic idea is solid!